### PR TITLE
Minor open-addressing key-list hashing changes

### PR DESCRIPTION
**Pull request by dalaro**
Hi @snazy.  Sorry I didn't review #4867 in time.  I saw the review ping, but I was too slow on the turnaround.

The code you wrote for open-addressing hashing of key-lists is clean and fascinating.  To understand the fix you pinged me about, I also looked the hash implementation generally (#4536) for background.  This is very nice.

#4867 looks good to me.  I may have found some minor tangential fixes to complement it.  I think there's basically one meaningful fix in here at most, which is described the first bullet point below.  The rest is either related to nomenclature consistency (bucket vs segment), testing or fixing extremely unlikely edge cases involving severe misconfiguration, or documentation and test additions.

I apologize for the size.  It's mostly tests and comments.  Furthermore, the only (intentional) substantive functionality changes are confined to `FetchValuesUsingOpenAddressing`, where the file-level diffstat is 60 lines.

_(With that, let me switch over to the PR branch's sole commit message, reproduced below)_

This commit makes a cluster of intertwined changes related to storage of keylists using open-addressing hashing:

* Fixes lookups of keys which hash to the highest-indexed segment of an  open-addressed key-list, but which physically reside in another   segment due to collisions (overwhelmingly likely to reside in the  embedded segment).  Also adds a unit test and integration test  exercising this case.  The unit test is  `TestFetchValuesUsingOpenAddressing#segmentWrapAroundWithPresentKey`,  which fails on `FetchValuesUsingOpenAddressing` from main (expect to  fix a method rename when mixing revs though: `getKeyListSegmentCount`  becomes `getKeyListBucketCount` in this commit).

* Attempts to make the word "segment" consistently refer to a collection  of contiguous hashtable cells, corresponding to either the embedded  keylist or a non-embedded keylist entity
* Attempts to make the word "bucket" consistently refer to a single cell  in the virtual keylist hashtable spread across all of the segments.  This extends to persist.proto, where I renamed a field, but did not  alter any tags, so it should remain wire-compatible.

* Made `FetchValuesUsingOpenAddressing#segmentForKey(bucket, round)` stop  clamping segment indices using the open addressing bucket index mask.  Instead, it sums the segment index and round, then applies the  remainder operator using the total segment count (embedded + not) as  the divisor.

* Added tests exercising some pathological edge cases.  Although these  elicited some failures against main, i.e. by checking out  `FetchValuesUsingOpenAddressing` from main, the importance is debatable.  Getting into these states generally requires major misconfiguration.  One case involving segment index overrun exercised by the new  `segmentWrapAround` test method might have come up, but the rest seem  low-probability.

* Made javadoc additions and changes on the read and write paths.

### CODE REVIEW

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 1]
```diff
@@ -90,17 +89,19 @@ List<Hash> entityIdsToFetch(int round, int prefetchEntities, Collection<Key> rem
...
           int entitySegment = segment - 1;
-          if (keyListIds.size() > entitySegment) {
-            Hash entityId = keyListIds.get(entitySegment);
-            if (entityIdToSegment.put(entityId, segment) == -1) {
-              entitiesToFetch.add(entityId);
-            }
+
+          Hash entityId = keyListIds.get(entitySegment);
+          if (entityIdToSegment.put(entityId, segment) == -1) {
+            entitiesToFetch.add(entityId);
```
Comment (dalaro) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:104]: I'm removing the `keyListIds.size() > entitySegment` condition because I'm now relying on `segmentForKey` to ensure that property (wrapping if necessary), and confirming it just above in `checkSegmentIndexValidity`.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 2]
```diff
@@ -58,28 +60,25 @@ final class FetchValuesUsingOpenAddressing {
...
   /**
    * Identifies the segment for a bucket. Segment 0 is the embedded key list, segment 1 is the first
    * key-list-entity.
    */
+  @VisibleForTesting
   int segmentForKey(int bucket, int round) {
     int binIdx = Arrays.binarySearch(keyListSegmentOffsets, bucket);
     int segment = binIdx >= 0 ? binIdx + 1 : -binIdx - 1;
-    return segment(segment) + round;
+    return (segment + round) % keyListsArray.length;
```
Comment (dalaro) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:81]: I'm very curious about this line and whether you think this looks valid, since I know it just went through an edit in #4867.  However, the old behavior of ANDing the segment index against the bucket index mask seemed like a no-op under sane configurations, and could actually trigger an erroneous miss in a pathologically bad configuration (exercised by one of the new ITs).

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 3]
```diff
@@ -163,4 +167,16 @@ Collection<Key> checkForKeys(
     }
     return keysForNextRound;
   }
+
+  private void checkSegmentIndexValidity(final int segment, final Key key) {
+    // We expect segmentForKey to wrap in case the current round number is large enough to overrun
+    // the segment list, so this shouldn't happen
+    if (keyListsArray.length <= segment) {
+      throw new IllegalStateException(
```
Comment (snazy) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:175]: Nit: `Preconditions.checkState` (provides string-formatting)

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 4]
```diff
@@ -58,28 +60,25 @@ final class FetchValuesUsingOpenAddressing {
...
   }
 
   /** Calculates the open-addressing bucket for a key. */
-  int segmentForKey(Key key) {
-    return segment(key.hashCode());
-  }
-
-  private int segment(int num) {
-    return num & segmentMask;
+  int bucketForKey(Key key) {
```
Comment (snazy) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:69]: Nice rename!

File: versioned/persist/serialize-proto/src/main/proto/persist.proto [hunk 5]
```diff
@@ -43,7 +43,7 @@ message CommitLogEntry {
   // Open-addressing bucket offsets. Same number of elements as key_list_ids.
   repeated int32 key_list_entity_offsets = 13;
   optional float key_list_load_factor = 14;
-  optional int32 key_list_segment_count = 15;
+  optional int32 key_list_bucket_count = 15;
```
Comment (snazy) [inline versioned/persist/serialize-proto/src/main/proto/persist.proto:46]: Nice rename!

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java [hunk 6]
```diff
@@ -154,6 +155,25 @@ static Stream<Arguments> openAddressing() {
...
+    // used for an int's shift distance, so the effective shift distance is 0.
+    assertThat(KeyListBuildState.nextPowerOfTwo(0)).isEqualTo(1);
+
+    assertThat(KeyListBuildState.nextPowerOfTwo(1)).isEqualTo(1);
+    assertThat(KeyListBuildState.nextPowerOfTwo((1 << 29) + 1)).isEqualTo(1 << 30);
+    assertThat(KeyListBuildState.nextPowerOfTwo((1 << 30) - 1)).isEqualTo(1 << 30);
+    assertThat(KeyListBuildState.nextPowerOfTwo(1 << 30)).isEqualTo(1 << 30);
+    assertThat(KeyListBuildState.nextPowerOfTwo((1 << 30) + 1)).isEqualTo(1 << 31);
+    assertThat(KeyListBuildState.nextPowerOfTwo(1 << 31)).isEqualTo(1 << 31);
+    // TODO might just want to throw an exception early in the following cases, in nextPowerOfTwo
```
Comment (snazy) [inline versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java:171]: Duh, yes! Please let it either throw an exception, before it becomes a negative value.

**Review by snazy — COMMENTED** [review]
Nice PR!
WIll do a another thorough review, but the changes look good and correct from a brief look.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java (hunk 2 above)
Comment (snazy) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:81]: Yea, you're right here - also your assumption "b/c of bad nomenclature" is also right.
`(segment + round) % keyListsArray.length` looks correct to me now - and this mod-operation is probably the reason why it's incorrectly been a mod-operation in the above methods above.
TL;DR very good catch!

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java (hunk 3 above)
Comment (snazy) [inline versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:175]: Is there any way we can easily trigger this in a test? No worries if not.

**Review by snazy — COMMENTED** [review]
The change looks really good!
Couple of nits (not fighting for those) and a few comments.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java (hunk 3 above)
Comment (dalaro) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java]: This was a thought-provoking comment.  The only ways I could induce this exception failure involve modifying this particular file.  After looking at it for a bit, I couldn't do it by limiting myself to only caller behavior, even with an arbitrarily bad/corrupt commit log entry.  Although I like defensive preconditions in general, what I've added here seems like overkill, since any potential violation of this invariant is likely to happen in the same (relatively compact) file where it is checked.  I've just deleted the precondition entirely, relying instead on `segmentForKey`'s new remainder operation with `keyListsArray.length` as divisor.  Though I can certainly bring it back the precondition, if you prefer.

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java [hunk 7]
```diff
@@ -90,17 +89,19 @@ List<Hash> entityIdsToFetch(int round, int prefetchEntities, Collection<Key> rem
     // Identify the key-list segments to fetch
     List<Hash> entitiesToFetch = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java:99]: Good idea.  I've added:
* a range clamp for `prefetchEntities` on the interval `[0, keyListCount-1]`
* an `IllegalArgumentException` when `prefetchEntities` is negative
* tests to exercise nonnegative cases (including "wraparound" `prefetchEntities` that exercises the clamp) and the exception

File: versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java [hunk 8]
```diff
@@ -343,4 +350,168 @@ void pointContentKeyLookups(
...
+   * int)} computing the wrong segment index for the final key, since it clamps a segment index to
+   * the bucket index interval. Applying the open-addressing bucket mask to the segment index is
+   * normally a no-op, except in the extreme edge case described above.
+   *
+   * <p>It was possible to induce this bug without setting the goal sizes all the way down to zero.
+   * It just required values small enough to generate a size-zero embedded key list and a series of
+   * size-one external key lists after that.
+   *
+   * <p>This case is covered with greater specificity in {@linkplain
+   * TestFetchValuesUsingOpenAddressing#segmentWrapAround}.
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractManyKeys.java]: I actually didn't realize IDEA would handle this case.  I saw that I couldn't reference on `TestFetchValuesUsingOpenAddressing` from `AbstractManyKeys` given the current dependency tree, then settled for a shorter name, but I see IDEA has no problem searching rest of the project index to find it correctly once I qualify it.  Thanks!

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java]: Agreed, check and parameterized test (`nextPowerOfTwoBucketCountException`) added with some negative values and values exceeding 2^30

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        .keyList(embeddedKeyList)
+        .keyListBucketCount(segmentCount * segmentSize)
+        .keyListsIds(keyListIds)
+        .hash(Hash.of("c0ffee"))
+        .build();
+  }
+
+  /** Convert an int to a hex string, left-padding with 0 until it has even length. */
+  private static String intToPaddedHexString(int i) {
+    String hexString = Integer.toHexString(i);
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: Agreed, that's simpler and more legible.  I've added a precondition to make sure the input is in unsigned byte range, and added that to the javadoc.  That's more than enough for all of my uses already.

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+   * various +1 and -1 offsets in play for nonexact matches. This just helped me prove that it was
+   * correct to begin with, and should localize our search if we accidentally break it in the
+   * future.
+   */
+  @Test
+  public void segmentForKey() {
+    FetchValuesUsingOpenAddressing fetch =
+        new FetchValuesUsingOpenAddressing(getCommitFixture(2, 4));
+
+    assertEquals(0, fetch.segmentForKey(0, 0));
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: Can do.  I've rolled up both this loop and the next one in the file, which had a similar comment.

After rolling up the implied loops in both of these two tests, the two tests are nearly identical.  The next natural step is to merge them into a single parameterized test, where the sole parameter is the size of the segments as measured in buckets (1 or 2).  I've also done that.

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+   * <p>This should only happen if a user misunderstands what the open-addressing-related config
+   * options actually do, or if they manage to serialize abnormally huge {@code CommitLogEntry}
+   * instances individually exceeding the segment size limit.
+   */
+  @Test
+  public void singletonSegment() {
+    FetchValuesUsingOpenAddressing fetch =
+        new FetchValuesUsingOpenAddressing(getCommitFixture(1, 4));
+
+    assertEquals(0, fetch.segmentForKey(0, 0));
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: Yep (merged into previous test per comment just above).

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 12]
```diff
@@ -0,0 +1,294 @@
...
+
+    // Look for a non-existent key.
+    // The fake segments here are all full, so this will probe through the entire hashtable.
+    // Expect the final segment in round 0, then no segments (implying embedded) in round 1,
+    // the first non-embedded segment in round 2, etc.
+    // This has the side-effect of adding entityIdToSegment mappings, which we'll need to call
+    // entityLoaded in the next part of this test.
+    Hash hashOfFinalSegment = Hash.of("03");
+    List<Key> singleton = ImmutableList.of(nonExistentKey);
+    assertEquals(ImmutableList.of(hashOfFinalSegment), fetch.entityIdsToFetch(0, 0, singleton));
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: Noted, thanks for the pointer.

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+    Key nonExistentKey =
+        new KeyWithConfigurableHash(ImmutableList.of("foo"), el -> Integer.MAX_VALUE);
+
+    // Look for a non-existent key.
+    // The fake segments here are all full, so this will probe through the entire hashtable.
+    // Expect the final segment in round 0, then no segments (implying embedded) in round 1,
+    // the first non-embedded segment in round 2, etc.
+    // This has the side-effect of adding entityIdToSegment mappings, which we'll need to call
+    // entityLoaded in the next part of this test.
+    Hash hashOfFinalSegment = Hash.of("03");
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: I've inlined both instances of this

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+    if (1 == hexString.length() % 2) {
+      hexString = "0" + hexString;
+    }
+    return hexString;
+  }
+
+  /**
+   * Builds a fake {@linkplain KeyListEntity}.
+   *
+   * <p>All keys appearing in this entity have acustom hashCode() that returns {@linkplain
```
[the remark this answers is not in the document]
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestFetchValuesUsingOpenAddressing.java]: Good catch

File: versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java (hunk 6 above)
Comment (dalaro) [reply versioned/persist/adapter/src/test/java/org/projectnessie/versioned/persist/adapter/spi/TestKeyListBuildState.java]: Sounds good.  I've set the range of valid inputs to `[0, 2^30]`, with a `checkArgument` precondition, javadoc, and test (the same new parameterized test that exercises a negative-parameter-value exception, commented on further down).

File: versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java (hunk 3 above)
Comment (dalaro) [reply versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java]: Agreed, I should use the variadic `checkState` overload instead of this verbose construct.  However, after reflecting on your comment about the testability of this exception elsewhere in the review, I've removed this check (with more on the thought process in that comment's thread).

**Review by snazy — APPROVED** [review]
LGTM!

**Comment by dalaro** [comment]
I've just discovered a latent bug fixed by this PR.  While doing unrelated work in #3832 which exercises open-addresing key-list hashing, I ran batteries of put-heavy tests against an old commit which did not include this PR's changes.

One of these test configurations could throw exceptions in the following form (again, this is effectively before the PR merged):

```
java.lang.ArrayIndexOutOfBoundsException: Index 1976 out of bounds for length 1976
        at org.projectnessie.versioned.persist.adapter.spi.FetchValuesUsingOpenAddressing.checkForKeys(FetchValuesUsingOpenAddressing.java:147)
        at org.projectnessie.versioned.persist.adapter.spi.AbstractDatabaseAdapter.fetchValuesHandleOpenAddressingKeyList(AbstractDatabaseAdapter.java:1553)
        at org.projectnessie.versioned.persist.adapter.spi.AbstractDatabaseAdapter.lambda$keysForCommitEntry$45(AbstractDatabaseAdapter.java:1370)
        at java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:271)
        at org.projectnessie.versioned.persist.adapter.spi.DatabaseAdapterUtil$1.lambda$tryAdvance$0(DatabaseAdapterUtil.java:206)
        at org.projectnessie.versioned.persist.adapter.spi.AbstractDatabaseAdapter$1.tryAdvance(AbstractDatabaseAdapter.java:1061)
        at org.projectnessie.versioned.persist.adapter.spi.DatabaseAdapterUtil$1.tryAdvance(DatabaseAdapterUtil.java:199)
        at java.base/java.util.Spliterator.forEachRemaining(Spliterator.java:326)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
        at org.projectnessie.versioned.persist.store.PersistVersionStore.lambda$commit$1(PersistVersionStore.java:211)
        at org.projectnessie.versioned.persist.store.PersistVersionStore.lambda$commit$2(PersistVersionStore.java:247)
        at org.projectnessie.versioned.persist.adapter.spi.AbstractDatabaseAdapter.commitAttempt(AbstractDatabaseAdapter.java:215)
        at org.projectnessie.versioned.persist.tx.TxDatabaseAdapter.lambda$commit$15(TxDatabaseAdapter.java:425)
        at org.projectnessie.versioned.persist.tx.TxDatabaseAdapter.opLoop(TxDatabaseAdapter.java:881)
        at org.projectnessie.versioned.persist.tx.TxDatabaseAdapter.commit(TxDatabaseAdapter.java:417)
        at org.projectnessie.versioned.persist.adapter.spi.TracingDatabaseAdapter.commit(TracingDatabaseAdapter.java:133)
        at org.projectnessie.versioned.persist.store.PersistVersionStore.commit(PersistVersionStore.java:253)
        at org.projectnessie.versioned.TracingVersionStore.lambda$commit$3(TracingVersionStore.java:107)
        at org.projectnessie.versioned.TracingVersionStore.callWithTwoExceptions(TracingVersionStore.java:352)
        at org.projectnessie.versioned.TracingVersionStore.commit(TracingVersionStore.java:101)
        at org.projectnessie.versioned.MetricsVersionStore.lambda$commit$1(MetricsVersionStore.java:88)
        at org.projectnessie.versioned.MetricsVersionStore.delegate2ExR(MetricsVersionStore.java:269)
        at org.projectnessie.versioned.MetricsVersionStore.commit(MetricsVersionStore.java:87)
        at org.projectnessie.services.impl.TreeApiImpl.commitMultipleOperations(TreeApiImpl.java:668)
        at org.projectnessie.services.impl.TreeApiImplWithAuthorization.commitMultipleOperations(TreeApiImplWithAuthorization.java:287)
        at org.projectnessie.services.rest.RestTreeResource.commitMultipleOperations(RestTreeResource.java:154)
        at org.projectnessie.services.rest.RestTreeResource_Subclass.commitMultipleOperations$$superforward1(Unknown Source)
        at org.projectnessie.services.rest.RestTreeResource_Subclass$$function$$1.apply(Unknown Source)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.proceed(AroundInvokeInvocationContext.java:53)
        at io.quarkus.hibernate.validator.runtime.interceptor.AbstractMethodValidationInterceptor.validateMethodInvocation(AbstractMethodValidationInterceptor.java:68)
        at io.quarkus.hibernate.validator.runtime.jaxrs.JaxrsEndPointValidationInterceptor.validateMethodInvocation(JaxrsEndPointValidationInterceptor.java:35)
        at io.quarkus.hibernate.validator.runtime.jaxrs.JaxrsEndPointValidationInterceptor_Bean.intercept(Unknown Source)
        at io.quarkus.arc.impl.InterceptorInvocation.invoke(InterceptorInvocation.java:41)
        at io.quarkus.arc.impl.AroundInvokeInvocationContext.perform(AroundInvokeInvocationContext.java:40)
        at io.quarkus.arc.impl.InvocationContexts.performAroundInvoke(InvocationContexts.java:32)
        at org.projectnessie.services.rest.RestTreeResource_Subclass.commitMultipleOperations(Unknown Source)
        at org.projectnessie.services.rest.RestTreeResource_ClientProxy.commitMultipleOperations(Unknown Source)
```

I had not seen this before.  I wanted to understand the root cause, just in case the implementation was still susceptible to it despite the changes here.  I reran tests with a debugger attached and caught this situation with a conditional breakpoint.  After tripping, I found this case involved a `hashCode()` triggering an edge case in what was [formerly called `segment(int)`](https://github.com/projectnessie/nessie/blob/36a22ed87e9e2c4f4df4fb6d5460f72a4111ea1d/versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java#L74):

```java
  private int segment(int num) {
    // This is compatible to 'num & (segmentCount-1)' (if segmentCount is a power of 2).
    int mod = num % segmentCount;
    return num >= 0 ? mod : segmentCount + mod;
  }
```

In the exception case, I had:

* `key.hashCode()` = -300679168
* `segmentCount`, which really described the number of buckets in the entire hash at this revision = 262144

This results in `mod = -300679168 % 262144 = 0`.  The problem is that we then return `segmentCount + 0 = segmentCount`.  The caller, `checkForKeys`, then treats this value as a 0-based indexed into the total virtual bucket space, causing it to run off the end by exactly one (accessing one index beyond the final array-element in the final segment/entity).

I think that this PR already fixed this case.  After merge, `checkForKeys` now calls [`bucketForKey`](https://github.com/projectnessie/nessie/blob/8efaf76f8ce1ceea9eebb9221fb2731e4143bdf1/versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/FetchValuesUsingOpenAddressing.java#L64-L71) to get this same index, and that uses bitwise-and against a mask generated by taking ((total bucket count in the entire hash) - 1).

**Comment by dalaro** [comment]
In closer examination, I think @snazy probably fixed this bug before I did, via https://github.com/projectnessie/nessie/pull/4867.  The version I found affected was older than his PR.  I haven't tried to reproduce the problem against 4867 specifically, but source reading suggests that 4867 should fix it, because that changeset replaced the problematic remainder operation with bitwise-and against a safe mask.
